### PR TITLE
Folding icon collides with git decorations: Create a div per linedecoration

### LINKED ISSUE #9222: [folding] missing outlining expander
- VSCode Version:
Version 1.3.1
Commit e6b4afa53e9c0f54edef1673de9001e9f0f547ae
Date 2016-07-12T13:35:06.227Z
Shell 0.37.6
Renderer 49.0.2623.75
Node 5.10.0
- OS Version:
Windows 10

Steps to Reproduce:

1.  Create a folder, and run git init inside it.

2.  Open the folder in Visual Studio Code.

3. Create a file with the following 6 lines, and commit:
var foo: number[] = [
    1,
];
var bar: number[] = [
    2,
]; // comment

4. Remove lines 3, 4, and 5 to leave:
var foo: number[] = [
    1,
]; // comment

5. Collapse the array using the minus to hide the contents of the array.

Result:
Notice the + expander is not present. The red marker seems to hide it.

Expected:
Collapsed sections always show the expander, otherwise.

vscode_bug_missingexpander

### PR DESCRIPTION
**Pull request by aeschli**
Motivated by issue #9222.  Pull request also contains the changes to fix #9222.

The underlying issue is that all line decorations on the same line share the same div and that css rules on that div interfere with each other. 

The fix creates a div per decoration className. The existing code already folds multiple decorations on the same line with the same class name.
The divs are still absolutely positioned, at the same location, so there's not more horizontal space used than before.

The same change could be done on the glyph margins, but I kept the existing behaviour.

Before
image

After
image

### CODE REVIEW

File: src/vs/editor/browser/viewParts/linesDecorations/linesDecorations.ts [hunk 1]
```diff
@@ -116,11 +116,7 @@ export class LinesDecorationsOverlay extends DedupOverlay {
 			if (classNames.length === 0) {
 				output[lineIndex] = '';
 			} else {
-				output[lineIndex] = (
-					'<div class="cldr'
-					+ classNames
-					+ common
-				);
+				output[lineIndex] = classNames.map(className => '<div class="cldr ' + className + common).join('');
```
Comment (alexandrudima) [inline src/vs/editor/browser/viewParts/linesDecorations/linesDecorations.ts]: please spell out for loops

File: src/vs/editor/contrib/folding/browser/folding.css [hunk 2]
```diff
@@ -6,40 +6,31 @@
 .monaco-editor .margin-view-overlays .folding {
 	margin-left: 5px;
 	cursor: pointer;
-}
-
-.monaco-editor .margin-view-overlays .folding::after {
 	background-repeat: no-repeat;
```
Comment (alexandrudima) [inline src/vs/editor/contrib/folding/browser/folding.css]: why are these changes part of this PR? Is the change in the editor core forcing these changes?

**Comment by alexandrudima** [comment]
@aeschli 

I forgot that I wrote the `DedupOverlay` that does a first round of deduping. The change LGTM. 

One nitpick about writing the for loop explicitly, because I prefer to spell out for loops in perf-sensitive rendering code. 

The other one is about the css change in the folding.css which seems unrelated to me.

**Comment by aeschli** [comment]
The css change is a separate commit (62ebdc3) and the fix for #9222, as the description says.
You don't need to review that.

**Comment by aeschli** [comment]
@alexandrudima I added a change set for the for loop

**Pull request merged**